### ISSUE TITLE
Problem trying to resolve a methodcall in a class with @Data annotation

### ISSUE THREAD

**Issue by Zou-Fei** [javaparser/javaparser#2436]
Class A:
```java
@Data
class ClassA {
    private String age;
}
```

`ClassB` with a method `getA` call the `getAge()` of `ClassA`:
```java
class ClassB {
    public void getA(){
        ClassA classA = new ClassA();
        classA.getAge();
    }
}
```
when I try to run the main class :
```java
public static void main(String[] args) throws Exception{
        String javaFile = "file/my/ClassB.java";

        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver();
        combinedTypeSolver.add(new ReflectionTypeSolver());
        combinedTypeSolver.add(new JavaParserTypeSolver("file/my"));
        JavaSymbolSolver symbolSolver = new JavaSymbolSolver(combinedTypeSolver);
        StaticJavaParser.getConfiguration().setSymbolResolver(symbolSolver);

        CompilationUnit cu = StaticJavaParser.parse(new FileInputStream(javaFile));
        List<String> methodCallList = new ArrayList<>();
        for(MethodDeclaration md: cu.findAll(MethodDeclaration.class)){
            md.accept(new MethodUnit(),methodCallList);
        }
    }
```

this is the error info:
> Exception in thread "main" UnsolvedSymbolException{context='null', name='We are unable to find the method declaration corresponding to classA.getAge()', cause='null'}

MethodUnit.java:
```java
public class MethodUnit extends VoidVisitorAdapter<List> {

    @Override
    public void visit(MethodCallExpr n, List args) {
        super.visit(n, args);
        ResolvedMethodDeclaration resolvedMethodDeclaration = n.resolve();
        args.add(resolvedMethodDeclaration.getQualifiedSignature());
    }
}
```

So there is an error with `@Data` annotation when run the resolve method, or did i make some mistakes?

**Comment by jlerbsc** [comment]
Could we assume that "file/my" is your root source directory (not a package directory)?

**Comment by Zou-Fei** [comment]
It's also the package directory, actually, when I removed the `@Data` annotation and add the `getAge()` method in `ClassA`, it can be successful, and when I debug the source code, I didn't find special handling of `@Data` when no find of getter or setter, did I miss something?

**Comment by jlerbsc** [comment]
Assuming that "file" is you root source files directory, you have to set JavaParserTypeSolver like that.

    combinedTypeSolver.add(new JavaParserTypeSolver("file"));

**Comment by Zou-Fei** [comment]
You can know that the "file/my" is the root source files directory, I can be sure that there is no question about the file directory, just maybe about "@Data"... when I tried to get all the method call in all class of my project, I find it will be failed to get the ResolvedMethodDeclaration of the getter method where the JavaBean with "@Data" annotation， **because there is no getter method in the static java file**.

**Comment by edefazio** [comment]
Just for clarity, can you explain where `@Data` comes from (what is it's fully qualified class name)

...If you are talking about the Project Lombok `@Data` annotation:
[https://projectlombok.org/features/Data](https://projectlombok.org/features/Data)

then I think I know what the problem is...

Lombok runs as a part of the annotation processing in the Compiler, so in order for that getter method to be generated and resolved using the `SymbolSolver`, you have to run the compiler FIRST... but in your example, you are adding the source path (pre-compilation) and then trying to have the symbol solver try to resolve the getter (but the .java source will still have the unprocessed @Data annotation as it will only get processed via Lomboks annotation processing)

Thats why it works when you manually add the getter... 

In short ... I don't think this is an issue with the SymbolSolver, but rather the expectation that you can put the source path on the Symbol Solver's Type Solver and expect it "understands" or processes Lomboks annotations (The Source Resolver of a `CombinedTypeSolver` doesnt compile/ process the annotations.)

**Comment by Zou-Fei** [comment]
Yes, it's Lombok @Data, so I totally understand what you mean，why I open this issue is to ask if there any solution to solve this problem, there are so many Lombok @Data in the project of my team, so when I trying to get all the method calling relationship of the whole project using SymbolSolver, I get so many errors because the Lombok @Data annotation, when solving  the getter method, or solving the method with getter method parameter..

**Comment by matozoid** [comment]
Random idea: it might be possible to extend a TypeSolver to fake the availability of the get and set methods? Or you can post-process the sources that the typesolver reads and add the getters and setters to the CU?

**Comment by edefazio** [comment]
To me the solution would be to include Lomboks ClassGeneration Path (i.e. the path it uses for where to write the Class files during annotation processing)

or Lomboks source generation path as a TypeSolver (AND REMOVE THE "ORIGINAL" SOURCE PATH) 
i.e.

This seems certainly like a Lombok issue and not a JavaParser/SymbolSolver issue as it deals with Lomboks lifecycle (of which I am not too familiar)

I'd do this:
1) when you are compiling and running the program (that is expecting the generated code) make sure it is set up with the appropriate switches for lombok (the annotation processing of Lombok is done... this will/should ALWAYS happen before RUNNING the code.)
2) making Sure Lombok IS writing out the source and you include this directory
3) ensure the Lombok "output directory" is added to the TypeSolver

In short 
I think the situation you are in now is that you have a compile-edit-run cycle that INCLUDES running Lombok annotation processing and then "testing" or whathaveyou

and another compile-edit-run setup that DOES NOT run Lombok's annotation processing during Javac compilation but is configured to interpret the raw (un-Lomboked) code expecting to have those methods generated

what you are looking for is possible but you are probably going to find the solution by having a better understanding of Lombok and updating your runtime environment.

My 2 cents
Eric

**Comment by edefazio** [comment]
...it would be nice to know the answer to this, so please report back on how you fixed it... (with details from the Lombok side)
(I'm not trying to be dismissive, but my Lombok understanding is minimal...)

I'd argue that the route above might be the best way of doing things...( but dont personally use Lombok...it'd be nice if someone who uses it can confirm or discredit what I put above)  

I'd hate to pin the responsibility on a specific TypeSolver to do the job as the SymbolSolver is already trying to do alot and that seems like feature creep/hack (sorry @matazoid )
 
I can certainly see this issue coming up again given the popularity of Lombok so do please let us know.

**Comment by matozoid** [comment]
@Zou-Fei - can you tell us how you fixed it?

**Comment by jlerbsc** [comment]
All the issues related to the use of lombok are widely commented on and the issues are closed. This can also be closed.
